src/hotspot/share/opto/regmask.hpp

```diff
@@ -65,27 +73,285 @@ class RegMask {
   static const unsigned int _WordBitMask = BitsPerWord - 1U;
   static const unsigned int _LogWordBits = LogBitsPerWord;
   static const unsigned int _RM_SIZE     = LP64_ONLY(RM_SIZE >> 1) NOT_LP64(RM_SIZE);
+  static const unsigned int _RM_SIZE_MIN =
+      LP64_ONLY(((RM_SIZE_MIN + 1) & ~1) >> 1) NOT_LP64(RM_SIZE_MIN);
   static const unsigned int _RM_MAX      = _RM_SIZE - 1U;
 
+  // Compute a best-effort (statically known) upper bound for register mask
+  // size in 32-bit words. When extending/growing register masks, we should
+  // never grow past this size.
+  static const unsigned int RM_SIZE_MAX =
+      (((RM_SIZE_MIN << 5) +                // Slots for machine registers
+        (max_method_parameter_length * 2) + // Slots for incoming arguments
+        (max_method_parameter_length * 2) + // Slots for outgoing arguments
+        BoxLockNode_slot_limit +            // Slots for locks
+        64                                  // Padding, reserved words, etc.
+        ) + 31) >> 5; // Number of bits -> number of 32-bit words
+  static const unsigned int _RM_SIZE_MAX =
+      LP64_ONLY(((RM_SIZE_MAX + 1) & ~1) >> 1) NOT_LP64(RM_SIZE_MAX);
+
+  // Sanity check
+  STATIC_ASSERT(RM_SIZE <= RM_SIZE_MAX);
+
+  // Ensure that register masks cannot grow beyond the point at which
+  // OptoRegPair can no longer index the whole mask.
+  STATIC_ASSERT(OptoRegPair::can_fit((RM_SIZE_MAX << 5) - 1));
+
   union {
-    // Array of Register Mask bits.  This array is large enough to cover
-    // all the machine registers and all parameters that need to be passed
-    // on the stack (stack registers) up to some interesting limit.  Methods
-    // that need more parameters will NOT be compiled.  On Intel, the limit
-    // is something like 90+ parameters.
+    // Array of Register Mask bits.  This array is large enough to cover all
+    // the machine registers and usually all parameters that need to be passed
+    // on the stack (stack registers) up to some interesting limit. On Intel,
+    // the limit is something like 90+ parameters.
```

Comment:
Here you fixed the comment, so probably the other one needs to be fixed too ;)

Reply:
Yes, thanks!